ultralytics/yolo/engine/exporter.py

```diff
@@ -401,7 +407,7 @@ def forward(self, x):
                 return cls, xywh * self.normalize  # confidence (3780, 80), coordinates (3780, 4)
 
         LOGGER.info(f'\n{prefix} starting export with coremltools {ct.__version__}...')
-        f = self.file.with_suffix('.mlmodel')
+        f = self.file.with_suffix('.mlpackage' if mlprogram else '.mlmodel')
 
         bias = [0.0, 0.0, 0.0]
         scale = 1 / 255
```

Comment:
Before the domain-specific models(self.model.task == 'classify' or self.model.task == 'detect') are created, we can add the assertion code after line 414 like this:

```python
mlprogram_support = self.model.task == 'classify' or self.model.task == 'detect'
if mlprogram_support:
  assert mlprogram f"Should specify the `"mlprogram"` as a `format`, not {format}"
```